### ISSUE TITLE
HTML polyglot - warning for missing lang attribute

### ISSUE THREAD

**Issue by JohnLukeBentley** [jgm/pandoc#3486]
It might be useful to list my suggested follow up tasks, for @jgm, from the pull request "Issue237 html5 polyglot: Make default.html5 polyglot markup conformant." #3473, given the PR is now merged.

With respect to a default value for a title my suggestion was:

1. In default.html5 keep title as is

       <title>$if(title-prefix)$$title-prefix$ – $endif$$pagetitle$</title>

2.  In the case where an author doesn't provide a value in the source document:
  a. From the HTML writer send a warning message to the command line; and
  b. From the HTML writer *don't* pass along a default value of "Untitled" to default.html5. Let the resulting html fail as invalid: with an empty title. I'll call this "2b-without-untitled".

It looks like we have 1 and 2 done. And you've gone ahead (perhaps partly based on my previously corrupting influence) and provided a different solution to 2b-without-untitled, I'll identify it as "2b-with-untitled". Namely you *are* sending through "Untitled". I'll leave it up to you to decide whether 2b-without-untitled, or 2b-with-untitled is the way to go.

With respect to the lang values:

3. In addition to having lang-values-default-to-blank I recommend that, when an
author does not provide a lang value, then upon on pandoc command execution
a warning message like the following be provided:

    > Polyglot markup stipulates that 'The root element SHOULD always specify
the language'. It is therefore recommended you specify a language value in
your source document. See
https://www.w3.org/International/articles/language-tags/ for valid
language values.

    You mentioned
     > Still need to add a warning about lang, but I think I'll explore a more general mechanism that also checks for the right format.

In general I get the sense you've got those pending tasks in hand and are well placed to make further decisions about them once you get your hands back on the code. But if you want me to act as a sounding board I'd be most happy to.

**Comment by jgm** [comment]
The only thing here that still needs to be done is some kind of warning about `lang`.

**Comment by JohnLukeBentley** [comment]
If you want to reject 2b-without-untitled in favour of 2b-with-untitled, that sounds right: just a `lang` warning to do.

**Comment by jgm** [comment]
I'm still not sure about issuing an actual warning, since it's SO common for people to leave out lang.  Maybe an INFO level message would be more appropriate.

**Comment by JohnLukeBentley** [comment]
That's an idea worth thinking about.

Is outputting an INFO versus a WARN effected by a user setting, like `--verbose`? Or is it just a conceptual difference?

It is common to leave the language out. A remainder of the specs ...

At the HTML5 level ... [the HTML5 spec, section "3.2.5.3. The lang and xml:lang attributes"]( https://www.w3.org/TR/html/dom.html#the-lang-and-xmllang-attributes),
> The lang attribute (in no namespace) specifies the primary language for the element's contents and for any of the element's attributes that contain text. Its value must be a valid BCP 47 language tag, or the empty string. Setting the attribute to the empty string indicates that the primary language is unknown.

At the Polyglot level https://www.w3.org/TR/html-polyglot/#language-attributes
> The root element SHOULD always specify the language

... So the author is merely encouraged to provide a `lang` value but not required to. We have the html output covered if they chose not to provide a `lang` value: `lang="" xml:lang=""` will be output.

I do imagine that authors in the document draft phase will want to leave out extraneous metadata, like a `lang`. That speaks in favour of both:

* Not impeding them (which a mere message need not do I presume); and 
* Prompting them if, as they move toward finalizing their source document, they want some sort of validity guide (the message from pandoc reminds them they SHOULD provide a lang value).

What are all the error message levels of Pandoc? If there was at least ERROR, WARN, INFO ... then leaving out a `lang` value would strike me as more of a WARN than an INFO. Imagining a conceptual correspondence between:
* INFO passes compilation and here is information about what was done.
* WARN -> SHOULD: passes compilation and note this is something you should do but currently don't.
* ERROR -> MUST: doesn't pass compilation because you must do this.

TL;DR: I'd favour a WARN over INFO.

**Comment by mb21** [comment]
Maybe the user should be able to set some default language or the system language should be taken or something?

Otherwise, I think the message should only be emitted under `--verbose`, otherwise lots of people are going to be fed up with it since they don't fancy adding a YAML metadata block to all their md docs just to add the language, especially if they're always writing in the same language anyway.

**Comment by jgm** [comment]
Hm, we could default to the setting of the `LANG` environment variable.

The difference between WARNING and INFO:  a warning will print by default (unless `--quiet`), and info only prints if you specify `--verbose`.  Also, you'll be able to specify that warnings should lead to exit with an error status (so you can have scripts ensure there are no warnings).  Given all this, I think INFO would be more appropriate.

**Comment by JohnLukeBentley** [comment]
Defaulting to a `lang` variable at the pandoc or OS level sounds like a good idea, @mb21. 

And I've noted the consequences of INFO V WARNING.

So with a defaulted `lang` variable and an INFO message (something like  ...

> Pandoc has defaulted to use "en-GB" as the language setting for you document. You can choose to explicitly override this at the Pandoc level by, ....; or you can choose to override this in your source document using the `lang` meta variable.

...) in a common case (without `--quiet` or `--verbose`; and before bothering with a `lang` meta variable in your source document) you'll end up with the correct language variable in the output html ... without lifting a finger. You could even rely on that mechanism to keep your meta variables to a few (e.g. `author`, `date`, `title`, `abstract`).

So all that sounds great.

**Comment by jgm** [comment]
I'm actually having second thoughts about including the empty lang attribute when no language is specified.

See https://www.w3.org/International/tests/repository/html5/the-lang-attribute/results-lang

Summary:

> An empty lang attribute value overrode an earlier lang attribute in all browsers.

> An empty value in a lang attribute killed any value from the HTTP declaration in IE, Chrome, Safari and Opera, but not in Firefox.

> An empty lang value on the html element killed any default value from the meta element, except (again) for Firefox.

Since a lang value might be specified elsewhere (e.g. as a default from the meta element or in the HTTP header), it seems better not to include the `lang=""`, since it would suppress this value specified earlier.

This is from 2014, so it's possible that things work differently now, but I wouldn't want to count on that unless told explicitly.

I'm actually now leaning towards a more austere approach to both title and lang:  leaving them empty if nothing is specified, and issuing a warning for title and an info message for lang.

**Comment by jgm** [comment]
I'm also having second thoughts on the idea of populating lang from the LANG environment variable.  Generally I prefer to avoid this kind of implicit behavior, especially since people often write web pages in a language other than their primary one.  (I imagine many Europeans, for example, maintain web pages in English; having the lang value set automatically to their home language may come as a surprise.)

**Comment by JohnLukeBentley** [comment]
1. Well, leaving a title value out (with pandoc warning), when the author doesn't supply a value, would be in line with my last recommendation. A recommendation made on the basis it letting the resulting html fail as invalid allows the author to both: temporarily ignore providing a title; but be alerted to the problem when they validate the HTML (e.g. at https://validator.w3.org/nu/).

   But I'll remind you of my previously mentioned alternative: default to the basename of the source file. E.g BirdsAndBees.md -> `<title>BirdsAndBees</title>`

2. On the lang value, when the author doesn't supply a value, you seem to be entertaining a choice between sending along:

       <html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">

   ... and ...

       <html xmlns="http://www.w3.org/1999/xhtml">

   ... in favour of the latter.

   > Since a lang value might be specified elsewhere (e.g. as a default from the meta element or in the HTTP header), it seems better not to include the lang="", since it would suppress this value specified earlier.

   If `lang=""` is stipulated to mean "unknown", which it is in the HTML5 spec(s), then it seems to me it is *desirable* that lang values set elsewhere (like the HTTP header) are suppressed.

   lang value setting in the meta tag seems to be deprecated. E.g. From https://wiki.whatwg.org/wiki/MetaExtensions
   > Redundant with the lang attribute on the html element. (Browsers pay attention to the lang attribute but not dc.language) 

   Moreover, I'll be inclined to go with what the HTML spec(s) say given that: they are more likely to embody conclusions from previous discussions on the matter (especially coming from a 2014 doc, as you flag); and the specs define a standard which browsers are meant to move toward (if you target current browser behaviour that might break next week if it moves towards standards compliance).

   And just to repeat the HTML 5 spec on the matter, but verify it's essentially the same from whatwg's point of view (last time I quoted W3C):

   https://developers.whatwg.org/content-models.html#attr-lang

   > The lang and xml:lang attributes

   > The lang attribute (in no namespace) specifies the primary language for the element's contents and for any of the element's attributes that contain text. Its value must be a valid BCP 47 language tag, or the empty string. Setting the attribute to the empty string indicates that the primary language is unknown. 

   That is, WHATWG also requires an empty string (rather than removing `lang="" xml:lang=""` altogether).

   So I recommend `<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">`

3. On whether to provide a default lang value or not (with an info message either way). I suppose it's a little bit arbitrary. Either option will preserve HTML validity (which, unlike in the title case, is probably a good thing) ... so HTML validity is not at stake.  Given that `lang="" xml:lang=""` has meaning, it means "unknown language", and an author may want to stipulate that meaning (against the encouragement of the specs to specify a particular value): I'm persuadable  to not providing a default value (but ensuring lang="" xml:lang="" gets output).

**Comment by jgm** [comment]
I've gone with:

- title: default to basename of first input file, or Untitled if from stdin.
- issue warning if title unspecified
- issue info level message if lang unspecified
- include empty lang attributes

**Comment by JohnLukeBentley** [comment]
Sounds good! Thanks very much for all that.
